lib/python-indent.coffee

```diff
@@ -42,28 +30,62 @@ module.exports = PythonIndent =
 
     # Get base variables
     row = editor.getCursorBufferPosition().row
-    previousLine = editor.buffer.lineForRow(row - 1)
-    return unless row
-
-    # Based on configuration, branch to corresponding logic
-    if (matchOpeningDelimiter = PythonIndent.openingDelimiterIndentRegex.exec previousLine) isnt null
-      PythonIndent.indentOnOpeningDelimiter(editor, row, previousLine, matchOpeningDelimiter)
-    else if PythonIndent.hangingIndentRegex.test previousLine
-      PythonIndent.indentHanging(editor, row, previousLine)
-    else if PythonIndent.openingDelimiterUnindentRegex.test previousLine
-      PythonIndent.unindentOnOpeningDelimiter(editor, row, previousLine)
-
-  indentOnOpeningDelimiter: (editor, row, previousLine, matchOpeningDelimiter) ->
-    # Get index of opening delimiter.
-    # The indent should be one whitespace character past this (sorry, no tabs)
-    indentColumn = previousLine.lastIndexOf(matchOpeningDelimiter[1]) + 1
+    col = editor.getCursorBufferPosition().column
+
+    # Parse the entire file up to the current point, keeping track of brackets
+    lines = editor.getTextInBufferRange([[0,0], [row, col]]).split('\n')
+    # at this point, the newline character has just been added,
+    # so remove the last element of lines, which will be the empty line
+    lines = lines.splice(0,lines.length-1)
+
+    # A stack of [row, col] pairs describing where open brackets are
+    parseOutput = PythonIndent.parseLines(lines)
+    openBracketStack = parseOutput.openBracketStack
+    closeBracketStack = parseOutput.closeBracketStack
+    shouldHang = parseOutput.shouldHang
+    lastFunctionRow = parseOutput.lastFunctionRow
+
+    if shouldHang
+        console.log('here')
+        PythonIndent.indentHanging(editor, row, editor.buffer.lineForRow(row - 1))
+        return
+
+    return unless openBracketStack.length or closeBracketStack.length
+
+    if not openBracketStack.length
+        # can assume closeBracketStack is not empty
+        lastClosedBracketLocations = closeBracketStack.pop()
+        if lastClosedBracketLocations[1] == row-1
```

Comment:
Ahhh...well I learned something new. I typically use `is`, so I guess
that's why I haven't seen errors :smile:

Dustin Speckhals
On Feb 25, 2016 8:43 PM, "Kevin Rose" <notifications@github.com> wrote:

> In lib/python-indent.coffee
> <https://github.com/DSpeckhals/python-indent/pull/9#discussion_r54195112>:
>
> > +    openBracketStack = parseOutput.openBracketStack
> > +    closeBracketStack = parseOutput.closeBracketStack
> > +    shouldHang = parseOutput.shouldHang
> > +    lastFunctionRow = parseOutput.lastFunctionRow
> > +
> > +    if shouldHang
> > +        console.log('here')
> > +        PythonIndent.indentHanging(editor, row, editor.buffer.lineForRow(row - 1))
> > +        return
> > +
> > +    return unless openBracketStack.length or closeBracketStack.length
> > +
> > +    if not openBracketStack.length
> > +        # can assume closeBracketStack is not empty
> > +        lastClosedBracketLocations = closeBracketStack.pop()
> > +        if lastClosedBracketLocations[1] == row-1
>
> Trying out some code on http://coffeescript.org/ seems to suggest that ==
> and is in coffeescript both get mapped directly to === in JavaScript, and
> === is an error.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DSpeckhals/python-indent/pull/9/files#r54195112>.
>